core/src/test/java/com/linecorp/armeria/common/ResponsePeekTest.java

```diff
@@ -48,25 +49,125 @@ void peekHeaders() {
     }
 
     @Test
-    void peekHeadersWithMapChain() {
+    void peekHeadersThrow() {
+        final HttpResponse res = HttpResponse.of(ResponseHeaders.of(HttpStatus.OK), HttpData.ofUtf8("foo"));
+
+        assertThatThrownBy(() -> res
+                .peekHeaders(headers -> {
+                    if (HttpStatus.OK.equals(headers.status())) {
+                        throw new IllegalStateException();
+                    }
+                })
+                .peekHeaders(headers -> peekCount.incrementAndGet())
+                .aggregate().join())
+                .isInstanceOf(CompletionException.class)
+                .hasCauseInstanceOf(IllegalStateException.class);
+        assertThat(peekCount).hasValue(0);
+    }
+
+    @Test
+    void peekData() {
+        final ResponseHeaders headers = ResponseHeaders.of(HttpStatus.OK);
+        final HttpResponse res = HttpResponse.of(headers,
+                                                 HttpData.ofUtf8("foo"),
+                                                 HttpData.ofUtf8("bar"));
+        final HttpResponse transformed = res
+                .peekData(data -> assertThat(data.toStringUtf8()).doesNotContain("\n"))
+                .mapData(data -> HttpData.ofUtf8(data.toStringUtf8() + '\n'))
+                .peekData(data -> assertThat(data.toStringUtf8()).contains("\n"));
+
+        final AggregatedHttpResponse aggregated = transformed.aggregate().join();
+        assertThat(aggregated.contentUtf8()).isEqualTo("foo\nbar\n");
+    }
+
+    @Test
+    void peekTrailers() {
+        final ResponseHeaders headers = ResponseHeaders.of(HttpStatus.OK);
+        final HttpResponse noTrailers = HttpResponse.of(headers, HttpData.ofUtf8("foo"));
+        final AtomicBoolean invoked = new AtomicBoolean();
+        final HttpResponse transformed1 = noTrailers
+                .peekTrailers(trailers -> peekCount.incrementAndGet())
+                .mapTrailers(trailers -> {
+                    invoked.set(true);
+                    return trailers;
+                });
+
+        final AggregatedHttpResponse aggregated1 = transformed1.aggregate().join();
+        assertThat(invoked).isFalse();
+        assertThat(peekCount).hasValue(0);
+        assertThat(aggregated1.headers().status()).isEqualTo(HttpStatus.OK);
+        assertThat(aggregated1.contentUtf8()).isEqualTo("foo");
+        assertThat(aggregated1.trailers().isEmpty()).isTrue();
+
+        final HttpResponse withTrailers = HttpResponse.of(headers,
+                                                          HttpData.ofUtf8("foo"),
+                                                          HttpHeaders.of("status", "13"));
```

Comment:
nit, what do you think of using `:status` to ensure that only trailers are peeked/mapped

```suggestion
                                                          HttpHeaders.of(":status", "13"));
```
or 
```suggestion
                                                          HttpHeaders.of(HttpHeaderNames.STATUS, "13"));
```

Reply:
I choose `HttpHeaders.of(HttpHeaderNames.STATUS, "13"));`!